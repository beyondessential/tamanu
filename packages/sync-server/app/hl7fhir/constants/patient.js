export const PATIENT_LINK_TYPES = {
  REPLACES: 'replaces',
  REPLACED_BY: 'replaced-by',
  SEE_ALSO: 'seealso',
};
