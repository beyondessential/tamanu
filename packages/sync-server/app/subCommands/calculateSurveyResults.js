import { Command } from 'commander';
import { checkJSONCriteria } from '@tamanu/shared/utils/criteria';

import { log } from 'shared/services/logging';

import { initDatabase } from '../database';

const SURVEY_RESPONSE_BATCH_SIZE = 1000;

/**
 * IMPORTANT: We have 4 other versions of this method:
 *
 * - mobile/App/ui/helpers/fields.ts
 * - desktop/app/utils/survey.js
 * - shared/src/utils/fields.js
 * - sync-server/app/subCommands/calculateSurveyResults.js
 *
 * So if there is an update to this method, please make the same update
 * in the other versions
 */
const checkVisibilityCriteria = (component, allComponentsFromQuery, values) => {
  const allComponents = allComponentsFromQuery.map(x => ({
    dataElement: {
      code: x.code,
      type: x.type,
    },
  }));

  try {
    return checkJSONCriteria(component.visibilityCriteria, allComponents, values);
  } catch (error) {
    log.error(`Error message: ${error}`);

    return false;
  }
};

const formatResultText = value =>
  // as percentage
  `${value.toFixed(0)}%`;

const getAnswerByCode = answerRows => {
  const answerByCode = {};
  answerRows.forEach(answerRow => {
    answerByCode[answerRow.code] = answerRow.answer;
  });
  return answerByCode;
};

const calculateSurveyResult = async (store, surveyResponseId, surveyComponents) => {
  const [answerRows] = await store.sequelize.query(
    `
    SELECT program_data_elements.code AS "code",
    survey_response_answers.body AS "answer"
    FROM survey_response_answers
    INNER JOIN program_data_elements
      ON program_data_elements.id = survey_response_answers.data_element_id
    WHERE survey_response_answers.response_id = :surveyResponseId
  `,
    {
      replacements: {
        surveyResponseId,
      },
    },
  );

  const answerByCode = getAnswerByCode(answerRows);
  const visibleResultComponents = surveyComponents
    .filter(component => component.type === 'Result')
    .filter(component => checkVisibilityCriteria(component, surveyComponents, answerByCode));

  // use the last visible component in the array
  const finalResultComponent = visibleResultComponents[visibleResultComponents.length - 1];
  if (!finalResultComponent) {
    // this survey does not have a result field
    return { result: 0, resultText: '' };
  }

  const rawValue = answerByCode[finalResultComponent.code];

  // invalid values just get empty results
  if (rawValue === undefined || rawValue === null || Number.isNaN(rawValue)) {
    return { result: 0, resultText: finalResultComponent.detail || '' };
  }

  // string values just get passed on directly
  if (typeof rawValue === 'string') {
    return { result: 0, resultText: rawValue };
  }

  // numeric data gets formatted
  return {
    result: rawValue,
    resultText: formatResultText(rawValue, finalResultComponent),
  };
};

const calculateSurveyResultsInBatch = async (
  store,
  surveyId,
  surveyComponents,
  batchSize,
  offset,
) => {
  const [surveyResponseRows] = await store.sequelize.query(
    `
      SELECT survey_responses.id,
      survey_responses.encounter_id AS "encounterId"
      FROM survey_responses
      WHERE survey_id = :surveyId
      ORDER BY survey_responses.id ASC
      LIMIT :batchSize
      OFFSET :offset
    `,
    {
      replacements: {
        surveyId,
        batchSize,
        offset: offset * batchSize,
      },
    },
  );

  for (const { id: surveyResponseId, encounterId } of surveyResponseRows) {
    const { resultText } = await calculateSurveyResult(store, surveyResponseId, surveyComponents);

    // Update the result text and updated_at
    await store.sequelize.query(
      `
      UPDATE survey_responses
      SET result_text = :resultText,
      updated_at = CURRENT_TIMESTAMP(3)
      WHERE id = :surveyResponseId
    `,
      {
        replacements: {
          resultText,
          surveyResponseId,
        },
      },
    );

    // Also update encounters.updated_at to trigger syncing down
    // inner survey responses to mobile / lan
    await store.sequelize.query(
      `
      UPDATE encounters
      SET updated_at = CURRENT_TIMESTAMP(3)
      WHERE id = :encounterId
    `,
      {
        replacements: {
          encounterId,
        },
      },
    );
    log.info(`Result generated for survey response with id ${surveyResponseId}`);
  }
};

/**
 * Due to an issue that none of the surveyResponse resultText was synced from mobile to sync-server,
 * this sub command can be used to rerun all the survey result calculations in sync-server.
 * Most of the code is copied pasted from tamanu-mobile
 * https://github.com/beyondessential/tamanu-mobile/blob/e81c87df3acb4518c808fbad399eec031a05e4c3/App/ui/helpers/fields.ts#L90
 * TODO: When we're sure that resultText are all syncing correctly, we can remove this sub command
 * @param {*} store
 * @param {*} options
 */
async function calculateSurveyResults() {
  const store = await initDatabase({ testMode: false });

  const [surveyRows] = await store.sequelize.query(`
    SELECT DISTINCT surveys.id
    FROM surveys
    INNER JOIN survey_screen_components
      ON surveys.id = survey_screen_components.survey_id
    INNER JOIN program_data_elements
      ON program_data_elements.id = survey_screen_components.data_element_id
    WHERE program_data_elements.type = 'Result';
  `);

  for (const { id: surveyId } of surveyRows) {
    const [surveyComponents] = await store.sequelize.query(
      `
        SELECT program_data_elements.code AS "code",
          program_data_elements.type AS "type",
          survey_screen_components.visibility_criteria AS "visibilityCriteria",
          survey_screen_components.detail AS "detail"
        FROM survey_screen_components
        INNER JOIN program_data_elements
          ON program_data_elements.id = survey_screen_components.data_element_id
        WHERE survey_id = :surveyId
        ORDER BY survey_screen_components.screen_index, survey_screen_components.component_index ASC;
    `,
      {
        replacements: {
          surveyId,
        },
      },
    );

    const [surveyResponseCountRows] = await store.sequelize.query(
      `
        SELECT COUNT(*)
        FROM survey_responses
        WHERE survey_id = :surveyId;
    `,
      {
        replacements: {
          surveyId,
        },
      },
    );
    const surveyResponseCount = surveyResponseCountRows[0].count;
    const batchCount = Math.ceil(surveyResponseCount / SURVEY_RESPONSE_BATCH_SIZE);

    // Run in batches to avoid OOM issue
    for (let i = 0; i < batchCount; i++) {
      await calculateSurveyResultsInBatch(
        store,
        surveyId,
        surveyComponents,
        SURVEY_RESPONSE_BATCH_SIZE,
        i,
      );
    }
  }

  log.info('Finished calculating survey results');
  process.exit(0);
}

export const calculateSurveyResultsCommand = new Command('calculateSurveyResults')
  .description('Recalculate all survey results (intended to fix a specific bug)')
  .action(calculateSurveyResults);
