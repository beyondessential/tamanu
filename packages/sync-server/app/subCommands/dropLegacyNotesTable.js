import { Command } from 'commander';
import { log } from 'shared/services/logging';
import { initDatabase } from '../database';

export async function dropLegacyNotesTable() {
  log.info('Delete legacy note_pages and note_items table...');

  const store = await initDatabase({ testMode: false });
  const { Note } = store.models;

  try {
    await Note.sequelize.query(`
      DROP TABLE note_items;
      DROP TABLE note_pages;
    `);

    log.info(`Sucessfully dropped note_pages and note_items table`);
    process.exit(0);
  } catch (error) {
    log.info(`Command failed: ${error.stack}\n`);
    process.exit(1);
  }
}

export const dropLegacyNotesTableCommand = new Command('dropLegacyNotesTable')
  .description('Drop legacy note_pages and note_items table')
  .action(dropLegacyNotesTable);
