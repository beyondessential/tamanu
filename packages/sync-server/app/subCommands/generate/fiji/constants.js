export const NUM_VILLAGES = 50;
export const NUM_EXAMINERS = 10;
export const NUM_FACILITIES = 20;
export const REF_ID_PREFIX = 'tamanuFijiGenerate';
