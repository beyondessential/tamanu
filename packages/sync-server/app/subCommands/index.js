export * from './apiKeys';
export * from './calculateSurveyResults';
export * from './fhir';
export * from './generate';
export * from './generateVaccineCertificate';
export * from './importReport';
export * from './loadSigner';
export * from './migrate';
export * from './migrateImagingRequestsToLocationGroups';
export * from './printRoutes';
export * from './provision';
export * from './report';
export * from './saveCertificateRequest';
export * from './serve';
export * from './serveAll';
export * from './settings';
export * from './shell';
export * from './tasks';
export * from './user';
export * from './migrateVitals';
export * from './generateInitialVitalLogs';
export * from './migrateChangelogNotesToEncounterHistory';
