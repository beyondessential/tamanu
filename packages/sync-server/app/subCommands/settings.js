/* eslint-disable no-console */

import { promises as fs } from 'fs';
import { Command } from 'commander';
import TOML from '@iarna/toml';
import { canonicalize } from 'json-canonicalize';
import { buildSettingsRecords } from 'shared/models/Setting';

// it does work, but eslint doesn't like it
// eslint-disable-next-line import/no-unresolved
import { parse as parseJiK } from '@bgotink/kdl/json';

import { initDatabase } from '../database';

export async function listSettings(filter = '', { facility } = {}) {
  const {
    models: { Setting },
  } = await initDatabase({ testMode: false });

  const settingsTree = await Setting.get(filter, facility);
  if (!settingsTree || Object.keys(settingsTree).length === 0) {
    return 'No settings found';
  }

  const settings = buildSettingsRecords(filter, settingsTree, facility);

  const lines = settings.map(({ facilityId, key }) =>
    facilityId ? `${key} (facility: ${facilityId})` : key,
  );
  lines.sort();
  return lines.join('\n');
}

export async function getSetting(key, { facility } = {}) {
  const {
    models: { Setting },
  } = await initDatabase({ testMode: false });

  const setting = await Setting.get(key, facility);
  if (setting === undefined) {
    return '(no setting found)';
  }

  return `value:\n${canonicalize(setting)}`;
}

export async function setSetting(key, value, { facility } = {}) {
  const {
    models: { Setting },
  } = await initDatabase({ testMode: false });

  const setting = await Setting.get(key, facility);
  const preValue =
    setting && JSON.stringify(setting) !== '{}'
      ? `current value:\n${canonicalize(setting)}\n`
      : 'no current value\n';

  const newValue = JSON.parse(value);
  await Setting.set(key, newValue, facility);
  return `${preValue}\nnew value set`;
}

export async function loadSettings(key, filepath, { facility, preview } = {}) {
  const {
    models: { Setting },
  } = await initDatabase({ testMode: false });

  if (key.length < 1) {
    throw new Error('Key must be specified');
  }

  const file = (await fs.readFile(filepath)).toString();
  let value;
  if (filepath.endsWith('.json')) {
    value = JSON.parse(file);
  } else if (filepath.endsWith('.toml')) {
    value = TOML.parse(file);
  } else if (filepath.endsWith('.kdl')) {
    value = parseJiK(file);
  } else {
    throw new Error('File format not supported');
  }

  if (preview) {
    return JSON.stringify(value, null, 2);
  }

  await Setting.set(key, value, facility);

  const currentValue = await Setting.get(key, facility);
  return JSON.stringify(currentValue, null, 2);
}

export const settingsCommand = new Command('settings')
  .description('Manage settings')
  .addCommand(
    new Command('list')
      .description('list all setting keys')
      .argument('[filter]', 'only output keys matching this')
      .option('--facility <facility>', 'ID of facility to scope to')
      .action(async (...args) =>
        console.log(`-------------------------\n${await listSettings(...args)}`),
      ),
  )
  .addCommand(
    new Command('get')
      .description('get a setting')
      .argument('<key>', 'key to retrieve')
      .option('--facility <facility>', 'ID of facility to scope to')
      .action(async (...args) =>
        console.log(`-------------------------\n${await getSetting(...args)}`),
      ),
  )
  .addCommand(
    new Command('set')
      .description('set a setting')
      .argument('<key>', 'key to create/update')
      .argument('<value>', 'value in JSON')
      .option('--facility <facility>', 'ID of facility to scope to')
      .action(async (...args) =>
        console.log(`-------------------------\n${await setSetting(...args)}`),
      ),
  )
  .addCommand(
    new Command('load')
      .description('load settings from a file')
      .argument('<key>', 'key to load to')
      .argument('<file>', 'JSON, TOML, or KDL file to load settings from')
      .option('--facility <facility>', 'ID of facility to scope to')
      .option('--preview', 'Print the settings that would be loaded in JSON')
      .action(async (...args) =>
        console.log(`-------------------------\n${await loadSettings(...args)}`),
      ),
  );
