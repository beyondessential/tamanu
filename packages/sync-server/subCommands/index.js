export { migrate } from './migrate';
export { report } from './report';
export { serve } from './serve';
export { setup } from './setup';
export { calculateSurveyResults } from './calculateSurveyResults';
