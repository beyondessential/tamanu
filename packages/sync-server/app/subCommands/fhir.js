import { Command } from 'commander';
import { QueryTypes } from 'sequelize';
import * as yup from 'yup';

import { FHIR_INTERACTIONS } from 'shared/constants';
import { log } from 'shared/services/logging';
import { resourcesThatCanDo } from 'shared/utils/fhir/resources';

import { ApplicationContext } from '../ApplicationContext';

const materialisableResources = resourcesThatCanDo(FHIR_INTERACTIONS.INTERNAL.MATERIALISE);

async function showStatus() {
  const app = await new ApplicationContext().init();

  for (const Resource of materialisableResources) {
    const count = await Resource.count();
    const latest =
      (
        await Resource.findOne({
          order: [['lastUpdated', 'DESC']],
        })
      )?.lastUpdated?.toISOString() || 'never';

    const upstreamCount = await Resource.UpstreamModel?.count();
    const upstreamLatest =
      (
        await Resource.UpstreamModel?.findOne({
          order: [['updatedAt', 'DESC']],
        })
      )?.updatedAt?.toISOString() || 'never';

    log.info(
      `${Resource.name}: ${count}/${upstreamCount} records/upstream, last updated ${latest}/${upstreamLatest}`,
    );
  }

  await app.close();
}

async function doRefresh(resource, { existing, missing, since }) {
  const app = await new ApplicationContext().init();

  if (resource.toLowerCase() === 'all') {
    for (const Resource of materialisableResources) {
      if (!Resource?.UpstreamModel) continue;
      await doRefresh(Resource.fhirName, { existing, missing, since });
    }
    return;
  }

  const Resource = materialisableResources.find(
    r => r.fhirName.toLowerCase() === resource.toLowerCase(),
  );
  if (!Resource) throw new Error(`No such FHIR Resource: ${resource}`);

  const recordsToDo = (
    await Resource.sequelize.query(
      `
      SELECT upstream.id AS id
      FROM ${Resource.UpstreamModel.tableName} upstream
      LEFT JOIN fhir.${Resource.tableName} downstream ON downstream.upstream_id = upstream.id
      WHERE true
        AND upstream.deleted_at IS NULL
        ${missing ? 'AND downstream.id IS NULL' : ''}
        ${existing ? 'AND downstream.id IS NOT NULL' : ''}
        ${since ? `AND upstream.updated_at > $1` : ''}
    `,
      { type: QueryTypes.SELECT, bind: since ? [since] : [] },
    )
  ).map(({ id }) => id);

  let done = 0;
  log.info(`Going to refresh ${recordsToDo.length} records...`);
  for (const upstreamId of recordsToDo) {
    await Resource.materialiseFromUpstream(upstreamId);
    done += 1;
    if (done % 100 === 0) log.info(`Refreshed ${done} out of ${recordsToDo.length}`);
  }

  log.info('Resolving upstream references...');
  await Resource.resolveUpstreams();

  log.info(`Done refreshing ${done} ${Resource.fhirName} records`);
  await app.close();
}

export const fhir = async ({ status, refresh, existing, missing, since }) => {
  if (status || !refresh) return showStatus();

  return doRefresh(
    refresh,
    yup
      .object({
        existing: yup.boolean().default(false),
        missing: yup.boolean().default(false),
        since: yup
          .date()
          .nullable()
          .default(null),
      })
      .validateSync({ existing, missing, since }),
  );
};

export const fhirCommand = new Command('fhir')
  .description('FHIR integration utilities')
  .option('--status', 'show status (default)')
  .option('--refresh <Resource>', 'refresh a FHIR Resource (use `all` to do refresh all resources)')
  .option('--existing', 'only refresh already-materialised resources, not missing ones')
  .option('--missing', 'only materialise missing resources, leave existing ones alone')
  .option('--since <date>', 'filter to source tables that have been updated since that date only')
  .action(fhir);
