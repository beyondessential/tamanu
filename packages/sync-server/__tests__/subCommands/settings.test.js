import { join } from 'path';
import { mkdtemp, writeFile, rmdir } from 'fs/promises';
import { tmpdir } from 'os';
import { fake } from 'shared/test-helpers/fake';
import { Op } from 'sequelize';
import { SETTINGS_SCOPES } from '@tamanu/constants';
import { listSettings, getSetting, setSetting, loadSettings } from '../../app/subCommands/settings';
import { createTestContext } from '../utilities';

describe('settings', () => {
  let ctx;
  const facility = '38ee9650-9b0c-440f-8b0c-c9526fba35ea';
  beforeAll(async () => {
    ctx = await createTestContext();
    const { Facility } = ctx.store.models;
    await Facility.create(fake(Facility, { id: facility }));
  });
  afterAll(() => ctx.close());

  beforeEach(async () => {
    const { Setting } = ctx.store.models;

    await Setting.destroy({
      where: {
        key: {
          [Op.like]: 'test%',
        },
      },
    });

    await Setting.set('test', {
      value: 'test',
      tree: {
        flower: ['flower', 'girl'],
        branch: {
          leaf: 'leaf',
        },
      },
    });

    await Setting.set(
      'test.tree.branch',
      {
        leaf: 'weed',
        root: 'root',
      },
      facility,
      SETTINGS_SCOPES.FACILITY,
    );
  });

  describe('list (global)', () => {
    it('shows all settings', async () => {
      const list = await listSettings();
      // Remove all lines not starting with test.
      // As full contents of settings is not guaranteed to be the same
      expect(list.replace(/\n?^(?!test.).*$\n?/gm, '')).toMatchSnapshot();
    });
    it('shows some settings with a filter', () =>
      expect(listSettings('test.tree')).resolves.toMatchSnapshot());

    it('shows no settings with a non-matching filter', () =>
      expect(listSettings('nothinghere')).resolves.toMatchSnapshot());
  });

  describe('list (facility)', () => {
    it('shows all settings', async () => {
      const list = await listSettings('', { facility });
      // Remove all lines not starting with test.
      // As full contents of settings is not guaranteed to be the same
      expect(list.replace(/\n?^(?!test.).*$\n?/gm, '')).toMatchSnapshot();
    });

    it('shows some settings with a filter', () =>
      expect(listSettings('test.tree', { facility })).resolves.toMatchSnapshot());

    it('shows no settings with a non-matching filter', () =>
      expect(listSettings('nothinghere', { facility })).resolves.toMatchSnapshot());
  });

  describe('get (global)', () => {
    it('retrieves a scalar', () => expect(getSetting('test.value')).resolves.toMatchSnapshot());

    it('retrieves an array', () =>
      expect(getSetting('test.tree.flower')).resolves.toMatchSnapshot());

    it('retrieves an object', () =>
      expect(getSetting('test.tree.branch')).resolves.toMatchSnapshot());

    it('retrieves a non-existent key', () =>
      expect(getSetting('nothinghere')).resolves.toMatchSnapshot());
  });

  describe('get (facility)', () => {
    it('retrieves a scalar', () =>
      expect(getSetting('test.value', { facility })).resolves.toMatchSnapshot());

    it('retrieves an array', () =>
      expect(getSetting('test.tree.flower', { facility })).resolves.toMatchSnapshot());

    it('retrieves an object', () =>
      expect(getSetting('test.tree.branch', { facility })).resolves.toMatchSnapshot());

    it('retrieves a non-existent key', () =>
      expect(getSetting('nothinghere', { facility })).resolves.toMatchSnapshot());
  });

  describe('set (global)', () => {
    it('sets a new scalar', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.enabled', 'true')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.enabled')).resolves.toBe(true);
    });

    it('overwrites a scalar', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.value', '"test2"')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.value')).resolves.toBe('test2');
    });

    it('sets a new array', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.list', '["groceries", "antiques"]')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.list')).resolves.toStrictEqual(['groceries', 'antiques']);
    });

    it('overwrites an array', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.tree.flower', '["tulips"]')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.tree.flower')).resolves.toStrictEqual(['tulips']);
    });

    it('sets a new object', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.dict', '{"broken": "eggs"}')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.dict')).resolves.toStrictEqual({ broken: 'eggs' });
    });

    it('overwrites an object', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.tree.branch', '{ "leaf": "bug" }')).resolves.toMatchSnapshot();
      await expect(Setting.get('test.tree.branch')).resolves.toStrictEqual({ leaf: 'bug' });
      await expect(Setting.get('test.tree')).resolves.toStrictEqual({
        flower: ['flower', 'girl'],
        branch: {
          leaf: 'bug',
        },
      });
    });
  });

  describe('set (facility)', () => {
    it('sets a new scalar', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.enabled', 'true', { facility })).resolves.toMatchSnapshot();
      await expect(Setting.get('test.enabled', facility)).resolves.toBe(true);
    });

    it('sets a new scalar over a global one', async () => {
      const { Setting } = ctx.store.models;
      await expect(setSetting('test.value', '"test2"', { facility })).resolves.toMatchSnapshot();
      await expect(Setting.get('test.value')).resolves.toBe('test');
      await expect(Setting.get('test.value', facility)).resolves.toBe('test2');
    });

    it('sets a new array', async () => {
      const { Setting } = ctx.store.models;
      await expect(
        setSetting('test.list', '["groceries", "antiques"]', { facility }),
      ).resolves.toMatchSnapshot();
      await expect(Setting.get('test.list', facility)).resolves.toStrictEqual([
        'groceries',
        'antiques',
      ]);
    });

    it('sets a new array over a global one', async () => {
      const { Setting } = ctx.store.models;
      await expect(
        setSetting('test.tree.flower', '["tulips"]', { facility }),
      ).resolves.toMatchSnapshot();
      await expect(Setting.get('test.tree.flower')).resolves.toStrictEqual(['flower', 'girl']);
      await expect(Setting.get('test.tree.flower', facility)).resolves.toStrictEqual(['tulips']);
    });

    it('sets a new object', async () => {
      const { Setting } = ctx.store.models;
      await expect(
        setSetting('test.dict', '{"broken": "eggs"}', { facility }),
      ).resolves.toMatchSnapshot();
      await expect(Setting.get('test.dict', facility)).resolves.toStrictEqual({ broken: 'eggs' });
    });

    it('sets a new object over a global one', async () => {
      const { Setting } = ctx.store.models;
      await expect(
        setSetting('test.tree.branch', '{ "leaf": "bug" }', { facility }),
      ).resolves.toMatchSnapshot();

      await expect(Setting.get('test.tree.branch', facility)).resolves.toStrictEqual({
        leaf: 'bug',
      });
      await expect(Setting.get('test.tree.branch')).resolves.toStrictEqual({ leaf: 'leaf' });

      await expect(Setting.get('test.tree', facility)).resolves.toStrictEqual({
        flower: ['flower', 'girl'],
        branch: {
          leaf: 'bug',
        },
      });
      await expect(Setting.get('test.tree')).resolves.toStrictEqual({
        flower: ['flower', 'girl'],
        branch: {
          leaf: 'leaf',
        },
      });
    });
  });

  describe('load', () => {
    let tempdir;
    beforeEach(async () => {
      tempdir = await mkdtemp(join(tmpdir(), 'settings-'));
    });
    afterEach(async () => {
      await rmdir(tempdir, { recursive: true });
    });

    describe('a json file', () => {
      it('to global namespace', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.json');
        await writeFile(file, JSON.stringify({ really: { nice: 'eyes' } }));

        await expect(loadSettings('test.json', file)).resolves.toMatchSnapshot();

        await expect(Setting.get('test.json')).resolves.toStrictEqual({ really: { nice: 'eyes' } });
      });

      it('to a facility', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.json');
        await writeFile(file, JSON.stringify({ really: { blue: 'eyes' } }));

        await expect(loadSettings('test.json', file, { facility })).resolves.toMatchSnapshot();

        await expect(Setting.get('test.json')).resolves.toBe(undefined);
        await expect(Setting.get('test.json', facility)).resolves.toStrictEqual({
          really: { blue: 'eyes' },
        });
      });

      it('preview only', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.json');
        await writeFile(file, JSON.stringify({ really: { dull: 'eyes' } }));

        await expect(loadSettings('test.json', file, { preview: true })).resolves.toMatchSnapshot();

        await expect(Setting.get('test.json')).resolves.toBe(undefined);
      });
    });

    describe('a KDL file', () => {
      it('to global namespace', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.kdl');
        await writeFile(
          file,
          `
          test.kdl {
            character "Marten Reed"
            firstAppearance 1
            human true
          }
          `,
        );

        await expect(loadSettings('test.kdl', file)).resolves.toMatchSnapshot();

        await expect(Setting.get('test.kdl')).resolves.toStrictEqual({
          character: 'Marten Reed',
          firstAppearance: 1,
          human: true,
        });
      });

      it('to a facility', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.kdl');
        await writeFile(
          file,
          `
          test.kdl {
            character "Pintsize"
            firstAppearance 1
            human false
          }
          `,
        );

        await expect(loadSettings('test.kdl', file, { facility })).resolves.toMatchSnapshot();

        await expect(Setting.get('test.kdl')).resolves.toBe(undefined);
        await expect(Setting.get('test.kdl', facility)).resolves.toStrictEqual({
          character: 'Pintsize',
          firstAppearance: 1,
          human: false,
        });
      });

      it('preview only', async () => {
        const { Setting } = ctx.store.models;
        const file = join(tempdir, 'test.kdl');
        await writeFile(
          file,
          `
          test.kdl {
            character "Claire Augustus"
            firstAppearance 2203
            human true
          }
          `,
        );

        await expect(loadSettings('test.kdl', file, { preview: true })).resolves.toMatchSnapshot();

        await expect(Setting.get('test.kdl')).resolves.toBe(undefined);
      });
    });
  });
});
