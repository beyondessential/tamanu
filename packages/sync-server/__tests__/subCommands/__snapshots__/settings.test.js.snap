// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings get (facility) retrieves a non-existent key 1`] = `"(no setting found)"`;

exports[`settings get (facility) retrieves a scalar 1`] = `
"value:
"test""
`;

exports[`settings get (facility) retrieves an array 1`] = `
"value:
["flower","girl"]"
`;

exports[`settings get (facility) retrieves an object 1`] = `
"value:
{"leaf":"weed","root":"root"}"
`;

exports[`settings get (global) retrieves a non-existent key 1`] = `"(no setting found)"`;

exports[`settings get (global) retrieves a scalar 1`] = `
"value:
"test""
`;

exports[`settings get (global) retrieves an array 1`] = `
"value:
["flower","girl"]"
`;

exports[`settings get (global) retrieves an object 1`] = `
"value:
{"leaf":"leaf"}"
`;

exports[`settings list (facility) shows no settings with a non-matching filter 1`] = `"No settings found"`;

exports[`settings list (facility) shows some settings with a filter 1`] = `
"test.tree.branch.leaf
test.tree.branch.root (facility only)
test.tree.flower"
`;

exports[`settings list (global) shows no settings with a non-matching filter 1`] = `"No settings found"`;

exports[`settings list (global) shows some settings with a filter 1`] = `
"test.tree.branch.leaf
test.tree.flower"
`;

exports[`settings load a KDL file preview only 1`] = `
"{
  "character": "Claire Augustus",
  "firstAppearance": 2203,
  "human": true
}"
`;

exports[`settings load a KDL file to a facility 1`] = `
"{
  "character": "Pintsize",
  "firstAppearance": 1,
  "human": false
}"
`;

exports[`settings load a KDL file to global namespace 1`] = `
"{
  "character": "Marten Reed",
  "firstAppearance": 1,
  "human": true
}"
`;

exports[`settings load a json file preview only 1`] = `
"{
  "really": {
    "dull": "eyes"
  }
}"
`;

exports[`settings load a json file to a facility 1`] = `
"{
  "really": {
    "blue": "eyes"
  }
}"
`;

exports[`settings load a json file to global namespace 1`] = `
"{
  "really": {
    "nice": "eyes"
  }
}"
`;

exports[`settings set (facility) sets a new array 1`] = `
"no current value

new value set"
`;

exports[`settings set (facility) sets a new array over a global one 1`] = `
"current value:
["flower","girl"]

new value set"
`;

exports[`settings set (facility) sets a new object 1`] = `
"no current value

new value set"
`;

exports[`settings set (facility) sets a new object over a global one 1`] = `
"current value:
{"leaf":"weed","root":"root"}

new value set"
`;

exports[`settings set (facility) sets a new scalar 1`] = `
"no current value

new value set"
`;

exports[`settings set (facility) sets a new scalar over a global one 1`] = `
"current value:
"test"

new value set"
`;

exports[`settings set (global) overwrites a scalar 1`] = `
"current value:
"test"

new value set"
`;

exports[`settings set (global) overwrites an array 1`] = `
"current value:
["flower","girl"]

new value set"
`;

exports[`settings set (global) overwrites an object 1`] = `
"current value:
{"leaf":"leaf"}

new value set"
`;

exports[`settings set (global) sets a new array 1`] = `
"no current value

new value set"
`;

exports[`settings set (global) sets a new object 1`] = `
"no current value

new value set"
`;

exports[`settings set (global) sets a new scalar 1`] = `
"no current value

new value set"
`;
